src/Runner.Worker/Handlers/CompositeActionHandler.cs

```diff
@@ -452,22 +449,45 @@ private async Task RunStepAsync(IStep step)
                 SetStepConclusion(step, Common.Util.TaskResultUtil.MergeTaskResults(step.ExecutionContext.Result, step.ExecutionContext.CommandResult.Value));
             }
 
+            if (step.ExecutionContext.Result == TaskResult.Failed)
+            {
+                var continueOnError = false;
+                try
+                {
+                    var templateEvaluator = step.ExecutionContext.ToPipelineTemplateEvaluator();
+                    continueOnError = templateEvaluator.EvaluateStepContinueOnError(step.ContinueOnError, step.ExecutionContext.ExpressionValues, step.ExecutionContext.ExpressionFunctions);
+                }
+                catch (Exception ex)
+                {
+                    Trace.Info("The step failed and an error occurred when attempting to determine whether to continue on error.");
+                    Trace.Error(ex);
+                    step.ExecutionContext.Error("The step failed and an error occurred when attempting to determine whether to continue on error.");
+                    step.ExecutionContext.Error(ex);
+                }
+
+                if (continueOnError)
+                {
+                    SetStepConclusion(step, TaskResult.Succeeded);
+                    Trace.Info($"Updated step result (continue on error)");
+                }
+            }
+
             Trace.Info($"Step result: {step.ExecutionContext.Result}");
             step.ExecutionContext.Debug($"Finished: {step.DisplayName}");
             step.ExecutionContext.PublishStepTelemetry();
         }
 
         private void SetStepConclusion(IStep step, TaskResult result)
         {
+            step.ExecutionContext.Outcome ??= result;
             step.ExecutionContext.Result = result;
             SetStepsContext(step);
         }
         private void SetStepsContext(IStep step)
         {
             if (!string.IsNullOrEmpty(step.ExecutionContext.ContextName) && !step.ExecutionContext.ContextName.StartsWith("__", StringComparison.Ordinal))
             {
-                // TODO: when we support continue on error, we may need to do logic here to change conclusion based on the continue on error result
-                step.ExecutionContext.Global.StepsContext.SetOutcome(step.ExecutionContext.ScopeName, step.ExecutionContext.ContextName, (step.ExecutionContext.Result ?? TaskResult.Succeeded).ToActionResult());
+                step.ExecutionContext.Global.StepsContext.SetOutcome(step.ExecutionContext.ScopeName, step.ExecutionContext.ContextName, (step.ExecutionContext.Outcome ?? TaskResult.Succeeded).ToActionResult());
```

Comment:
can we do `step.ExecutionContext.Outcome ?? step.ExecutionContext.Result ?? TaskResult.Succeeded`

Reply:
Perhaps let's refactor `SetStepsContext` to be a part of `ExecutionContext` while we're at it?

Reply:
Sure!